modules/fx/arrow-fx/src/main/kotlin/arrow/fx/IOConnection.kt

```diff
@@ -1,48 +1,98 @@
-@file:Suppress("UnusedImports")
-
 package arrow.fx
 
-import arrow.core.Either
+import arrow.core.identity
+import arrow.fx.internal.JavaCancellationException
 import arrow.fx.typeclasses.Disposable
-import arrow.fx.typeclasses.ExitCase
-import arrow.fx.typeclasses.MonadDefer
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
 import kotlin.coroutines.AbstractCoroutineContextElement
 import kotlin.coroutines.CoroutineContext
-import arrow.fx.handleErrorWith as handleErrorW
 
-fun IOConnection.toDisposable(): Disposable = { cancel().fix().unsafeRunSync() }
-typealias IOConnection = KindConnection<ForIO>
+enum class OnCancel {
+  ThrowCancellationException, Silent;
+
+  companion object {
+    val CancellationException = ConnectionCancellationException
+  }
+}
+
+object ConnectionCancellationException : JavaCancellationException("User cancellation")
```

Comment:
Could we double check that these can be compared to themselves?

Reply:
You mean add a test that checks for reference checking?

```
//IOTest.kt

"ConnectionCancellationException can be reference checked" {
  ConnectionCancellationException shouldBe ConnectionCancellationException
}
```

Reply:
Yes, something like that :D `object` have comparison overriden anyway, right? There was some edge case where you could instantiate a second one after a rotation or something IIRC.

Reply:
Yes, I think it's possible to run in a weird double instance scenario but thought that was only possible in Android user land. When you have nested objects or inherit from certain lifecycle aware classes? 🤔

They should never be able to break a library object like `ConnectionCancellationException`, right? That'd be very scary if that'd be possible by accident.